wifiphisher/common/extensions.py

```diff
@@ -0,0 +1,305 @@
+        :rtype: None
+        """
+
+        self._extensions_str = extensions
+
+    def init_extensions(self, shared_data):
+        """
+        Init EM extensions. Should be run
+        when all shared data has been gathered.
+
+        :param self: An ExtensionManager object
+        :type self: ExtensionManager
+        :param shared_data: Dictionary object
+        :type shared_data: Dictionary
+        :return: None
+        :rtype: None
+        """
+
+        # Convert shared_data from dict to named tuple
+        shared_data = collections.namedtuple('GenericDict',
+                                             shared_data.keys())(**shared_data)
+        # Initialize all extensions with the shared data
+        for extension in self._extensions_str:
+            mod = importlib.import_module(
+                constants.EXTENSIONS_LOADPATH + extension)
+            ExtensionClass = getattr(mod, extension.title())
+            obj = ExtensionClass(shared_data)
+            self._extensions.append(obj)
+
+    def start_extensions(self):
+        """
+        Starts the two main daemons of EM:
+
+        1) Daemon that listens to every packet and
+        forwards it to each extension for further processing.
+        2) Daemon that receives special-crafted packets
+        from extensions and broadcasts them in the air.
+
+        :param self: An ExtensionManager object
+        :type self: ExtensionManager
+        :return: None
+        :rtype: None
+        """
+
+        # One daemon is listening for packets...
+        self._listen_thread.start()
+        # ...another daemon is sending packets
+        self._send_thread.start()
+        # daemon for channel hopping
+        self.get_channels()
+        self._channelhop_thread.start()
+
+    def on_exit(self):
+        """
+        Stops both daemons of EM on exit.
+
+        :param self: An ExtensionManager object
+        :type self: ExtensionManager
+        :return: None
+        :rtype: None
+        """
+
+        self._should_continue = False
+        if self._listen_thread.is_alive():
+            self._listen_thread.join(3)
+        if self._send_thread.is_alive():
+            self._send_thread.join(3)
+        if self._channelhop_thread.is_alive():
+            self._send_thread.join(3)
+
+    def get_channels(self):
+        """
+        Gets the channels from each extension.
+        Merges them to create a list of channels
+        to hop.
+
+        :param self: An ExtensionManager object
+        :type self: ExtensionManager
+        :return: None
+        :rtype: None
+        """
+
+        for extension in self._extensions:
+            channels_interested = extension.send_channels()
+            if channels_interested and len(channels_interested) > 0:
+                # Append only new channels (no duplicates)
+                self._channels_to_hop += list(set(channels_interested) -
+                                              set(self._channels_to_hop))
+
```

Comment:
Although it is more compact compared to the extended version I think it's harder to read at the first glance but that's just me.